### ISSUE TITLE
BERT: Pre-training of Deep Bidirectional Transformers for Language Understanding

### ISSUE THREAD

**Issue by Shoichi-Hasegawa0628** [Shoichi-Hasegawa0628/summary_paper#52]
https://arxiv.org/abs/1810.04805
https://qiita.com/omiita/items/72998858efc19a368e50

**Comment by Shoichi-Hasegawa0628** [comment]
- TransformerのEncoderを使用したモデル  

- Transformerを双方向に接続したモデル (ELMo (多層LSTMの双方向)のTransformer版)、注目する単語の前後両方の文脈について学習

- Fine-tunning based approach
事前学習で得たパラメータ (層間を繋ぐところの重み)を初期値として設定し、学習 (Fine-tunning)

**Comment by Shoichi-Hasegawa0628** [comment]
**入力形式**

- 入出力：1 or 2 sentence
　- 1 sentence → 感情分析など
　- 2 sentence → QAなど. 二文をくっつけて、入出力

- 入力形式
　- 先頭示すトークン：[CSL]
　- (二文を接続するとき)文の間 or 文の末尾を示すトークン：[SEP]
　
↓
文　　　　：今日の天気は雪だった。
トークン列：'[CSL]',  '今日', 'の' , '天気',  'は',  '雪',  'だっ', 'た', '。', '[SEP]'

**Comment by Shoichi-Hasegawa0628** [comment]
**ベクトル化**
トークン列にしたあと、
それぞれのトークンをベクトルに変換し、BERTに入力
(それぞれのトークンのベクトルをm次元に変換するものとする)

- トークン
(語彙数, m)サイズの行列E^Tを用意  
語彙中j番目のトークンが現れたら、E^Tのj行目の行ベクトルに置換  
→ 文章中のi番目のトークンをベクトルに置換したものを(e^T)_i

- 文章のタイプ
(2, m)サイズの行列E^Sを用意  
BERTに入力される文が1文：それぞれのトークンをE^sの1行目のベクトルに置換  
BERTに入力される文が2文：最初の文に含まれるトークンは1行目、次の文は2行目のベクトルに置換
→ 文章中のi番目のトークンをベクトルに置換したものを(e^S)_i

- 文章中の位置
BERTに入力可能な最大トークン数をLとし、(L, m)サイズの行列E^Pを用意  
→ 文章中のi番目のトークンをE^Pのi行目のベクトルに置換, (e^P)_i


よって文章中のi番目のトークンのベクトルは、
e_i = (e^T)_i + (e^S)_i + (e^P)_i

これをすべてのトークンに対して適用し、BERTにぶちこむ  
E^T, E^S, E^Pは学習されるパラメータ

**Comment by Shoichi-Hasegawa0628** [comment]
_**事前学習**_

大規模な文章コーパスを使い、汎用的な言語パターンを学習するために行う。
事前学習で用いるのは、**生の文章データ (ラベルなしデータ)**  

「モデルに何を学習させるか」がラベル有りと比較すると自明ではないので  
(最終的には誤差関数で正解との誤差出して、誤差逆伝搬するので工夫がいる)、
BERTは、以下の２つのタスクを使用
- **マスク付き言語モデル**
- **Next Sentence Prediction**

**Comment by Shoichi-Hasegawa0628** [comment]
_**マスク付き言語モデル**_
一部のトークンが[MASK]である文章から、[MASK]に入るトークンを予測する (正確には、それぞれのトークンに確率値を割り当てる) モデル  

おおまかな流れ

1. 全体のトークンからランダムに15%のトークンを選択
2. 選ばれたトークンに以下を適用
　- 80%を[MASK]トークン
　- 10%をランダムに選んだ他のトークン
　- 10%はそのまま 
3. トークン列をベクトルに変換
4. BERTに入力し、[MASK]トークンに対応するBERTの出力ベクトルを分類器 (多層MLPとか？)に入れ、予測確率を出す

→
[MASK]に置換されたトークンを、[MASK]を含む文章に対するラベルとして、学習を行う

**Comment by Shoichi-Hasegawa0628** [comment]
_**Next Sentence Prediction**_  
与えられた2文が連続化どうかを判定させるタスク  

事前学習のおおまかな流れ
1. 以下のデータを用意
　- データの50%：２つ目の文は1文目に連続
　- 残りの50%：２つ目の文はランダムに選ばれている
　- [CLS] に対するBERTの出力ベクトルを分類器に入れ、連続かどうかを判定

**Comment by Shoichi-Hasegawa0628** [comment]
**ファインチューニング**  
事前学習で得たパラメータを初期値とし、タスク用のラベル付きデータで学習を行わせること。


モデルのパラメータ (重み)の初期値を事前学習で得られたパラメータを使用し、新たに加えられた分類機のパラメータにはランダムな値を割り振る。

タスク用のラベル付きデータを使い、BERTも分類器も学習を行わせる。

**Comment by Shoichi-Hasegawa0628** [comment]
**事前学習、ファインチューニング、転移学習、ドメイン適応の用語整理**

**Comment by Shoichi-Hasegawa0628** [comment]
参考サイト(凄くわかりやすい)：
https://jinbeizame.hateblo.jp/entry/kelpnet_transfer

**事前学習 (Pre-Training)**  
大規模な文章コーパスを使い、汎用的な言語パターンを学習するために行う。
事前学習で用いるのは、生の文章データ (ラベルなしデータ)  
→ BERTやGPT



**ファインチューニング (Fine-Tuning)**  
事前学習で得た層間の重みを初期値とし、学習済みモデルの一部もしくは全てをタスク用のラベル付きデータで学習させること.
→ RoboCupとかで使うYOLO  
(重みは固定し、タスク用のデータセット食わせて、出力層だけ学習)



**転移学習 (Transfer Learning)**  
学習済みのモデルの重みは固定し、出力層 (新たに追加した分類器など)のみを学習させること。
→ ELMo
→→ (注意) 田口さんの「場所概念に基づく環境間の知識転移モデル」は転移学習ではない。  
https://www.tandfonline.com/doi/full/10.1080/01691864.2021.2004224
学習アルゴリズムを見ると、
環境数Eが(既知環境16 (画像・位置・言語) + 新規環境1 (画像・位置))で構成され、それをGibbs Samplingしてる。
つまり、新規環境における言語情報が欠損しているときに、他の環境で学習した言語情報の確率分布のパラメータを使うことで、場所の単語の予測を行う。
学習自体はバッチ学習なので、知識転移が確かに正しい.



**ドメイン適応 (Domain Adaptation)**  
出力層側の重みは固定し、入力層に近い層をタスク用のラベル付きデータで再学習すること。
(Ex. 音声認識
音声 → 単語 → 文章というように処理するが、単語→ 文章の出力層側の方法は共通だが、音声→単語は話者によって発音が異なるためここだけ学習)
![差異](https://user-images.githubusercontent.com/74911522/206893927-148018f5-e1d0-490b-9e1c-4b0fabd07a36.svg)
